torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,227 @@
+    if (isMutatingOp(node)) {
+      node->replaceInput(0, curToken);
+      return node->outputs().at(0);
+    }
+
+    JIT_ASSERT(node->blocks().size() == 0);
+
+    // For pure ops that need to be annotated, fence them.
+    if (shouldAnnotate(node)) {
+      return addFenceForNode(node, curToken);
+    }
+
+    return curToken;
+  }
+
+  bool shouldAnnotate(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (input->type()->kind() != TypeKind::ListType) {
+        // TODO(suo): Right now, we only support mutable lists.
+        // If we remove this check, it's not clear whether:
+        //
+        //   append(int[] a, int b)
+        //
+        // mutates `a` or `b`. We'll need to extend the schema language to be
+        // able to express which argument is mutated.
+        continue;
+      }
+      // First check the cache
+      if (mutableValues_.count(input) != 0) {
+        return true;
+      }
+
+      // Check whether any mutating op uses this input
+      for (const auto& use : input->uses()) {
+        if (isMutatingOp(use.user)) {
+          mutableValues_.insert(input);
+          return true;
+        }
+      }
+    }
+
+    // Check that any sub-blocks need to be annotated
+    for (auto block : node->blocks()) {
+      if (shouldAnnotate(block)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool shouldAnnotate(Block* block) {
+    for (auto node : block->nodes()) {
+      if (shouldAnnotate(node)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool isMutatingOp(const Node* node) {
+    return std::any_of(
+        node->inputs().begin(), node->inputs().end(), [](const Value* input) {
+          return input->type() == WorldType::get();
+        });
+  }
+
+  // Create a memory fence around a node, using the world token
+  //
+  // Input:
+  //  %size : Int = prim::len(%mut_list)
+  //
+  // Output:
+  //  %t.1 : World, %list.2 : int[] = prim::MemoryFence(%curToken, %mut_list)
+  //  %size : Int = prim::len(%mut_list)
+  //  %t.2 : World, %size.2 : int = prim::MemoryFence(%t.1, %size)
+  //
+  // Returns the new world token (%t.2) for subsequent fences to use.
+  Value* addFenceForNode(Node* node, Value* curToken) {
+    auto mut = node->inputs().at(0);
+
+    // Add a start fence
+    auto startFence =
+        node->owningGraph()->create(prim::MemoryFence, /*outputs=*/2);
+
+    startFence->addInput(curToken);
+    startFence->addInput(mut);
+
+    curToken = startFence->outputs()[0]->setType(WorldType::get());
+    mut = startFence->outputs()[1]->setType(mut->type());
+
+    startFence->insertBefore(node);
+
+    // modify the node to take in the start fence's output list
+    node->replaceInput(0, mut);
+
+    // Add an end fence
+    auto endFence =
+        node->owningGraph()->create(prim::MemoryFence, /*outputs=*/2);
+
+    endFence->addInput(curToken);
+    endFence->addInput(node->outputs().at(0));
```

Comment:
Also here there is no guarantee that there is only one output of the original node. All values should go through the fence.

Reply:
We should also re-wire all uses of the outputs to be the outputs after then fence, in order to prevent unexpected cases where other ops can get scheduled inside the fence operator.

Reply:
Please fence _all_ inputs and _all_ outputs. The reason for doing this is to maintain fiction that we really just have two versions of every op: one that takes control dependencies and one that doesn't. Fencing only some of the inputs and outputs allows some re-writes that break that invariant.  At the very least it will make it harder to reason about what might happen. Later when we make mutability a property of the type, we can be much looser about when memory fences get inserted and how world tokens are threaded. (e.g. we can allow reads of mutable values to happen in any order and only serialize writes).